### DISCUSSION TITLE
Writeup about things Taskwarrior would do differently in full rewrite

### DISCUSSION THREAD

**Discussion by dzintars** [GothenburgBitFactory/taskwarrior#3007] [Ideas]
I know this most likely will be unpopular idea, but still...
No software is perfect.
Over time we realize that we took not the best paths to build it.
It would be beneficial for the community to learn from the other mistakes.
What would be those things you would do differently if you would decide to 
write Taskwarrior from scratch?
For example:
- What you would do to appeal the community?
- What stack you would use?
- What architecture you would pick?
- What core features you would implement?
- How would you handle extensibility?
- etc, etc, etc...
- How ideal Taskwarrior would look tomorrow?

**Comment by djmitche** [comment]
I actually tried to rewrite Taskwarrior from scratch -- https://github.com/taskchampion/taskchampion.  There were reasons not to do so that have been discussed other places, but it illustrates at least some of my answers to these questions.

At the moment, Taskwarrior's in a bit of a slow spot -- there just aren't enough devs working on it to make measurable progress (one dev can't review their own PRs!).  So I think your first question is probably the most important: how would we get a "critical mass" of people involved?

**Comment by Matt-Deacalion** [comment]
I haven't been using Taskwarrior for long, but I really like it. There's a lot thought in it's design. For example: being able to write hooks in any language is very useful.

The main obstruction for me contributing is `C++`, it's not a language I know and Taskwarrior is the only project I would dedicate time to that's written in it - which means most of the time I could commit to it would be spent learning C++.

The parts that work well with other languages (or language-agnostic) have been really useful - hooks, Python unit tests, tasklib, rfcs, docs. It would be good if Taskwarrior provided a core library that could be used to build `task/list/workflow-like` processes and tools on-top of, people are so diverse in how they do things that I can't imagine a single tool satisfying everyone. I'm not sure how this would work - REST API, FFI, embedded Lua/JavaScript etc. but I imagine if the core team could focus on RFCs, docs and core functionality, there might be more bandwidth to move things forward.

**Reply by djmitche** [reply]
> It would be good if Taskwarrior provided a core library that could be used to build task/list/workflow-like processes and tools on-top of

That's a big part of the motivation for Taskchampion!

**Reply by Matt-Deacalion** [reply]
> > It would be good if Taskwarrior provided a core library that could be used to build task/list/workflow-like processes and tools on-top of
> 
> That's a big part of the motivation for Taskchampion!

I didn't realise that's what Taskchampion was! I'll take a look.

**Comment by dzintars** [comment]
Oh. Neat. I see Taskchampion use Rust. That is something I am really keen to learn in near near future. Unfortunately C is not on my radar.
I think Taskwarrior fits really well in Vim/Neovim ecosystem as it is where i do come from. I was looking for keyboard centric and super simple to create task tool to list my "quick list tasks". I mean, quickly drop in the list with ideas or tasks i should handle. And with option later augment my tasks with some metadata. Taskwarrior fits perfectly for what I was looking for.
Currently I am just at entry level "discovery" phase using it as my daily driver. I can't say about any real flaws which stops me. The only thing is, I would love to have something similar to `kdheepak/lazygit.nvim` UI, but to manage tasks. Switching Tmux sessions/windows is not a huge deal, but... i would love to have modal UI over my Neovim for less context switching and faster access when shaping the idea.
The other thing i can see is integration with Zettelkasten type note taking. I am using `zk` and `zk-nvim` which supports LSP. So, linking larger files to tasks and tasks to notes, would be "nice-to-have" (I saw there is a plugin which i should try).
I see there are more plugins around, but that is something I still need to explore closer.
But, back to main point. My personal opinion is that we need to hook into hot Neovim/Lua topic.
If there would be fancy new Neovim pugin to neatly manage tasks and to see fancy dashboards, i think we could get some traction there. We can use the same Taskwarrior as is. It is not bad at all. I am considering to try to write my first Neovim plugin in Lua and Taskwarrior is something that i see worth for it.
Before Taskwarrior I was using Stashpad for some time. But Electron killed that. It's just too much of context switch for me. But i liked idea of sub-tasking and markdown in the task itself. Thou... i should admit, that direct markdown support leads more into note taking than task management. :) I thing Taskwarriors annotations suits much better as it helps to focus on getting thing done.

**Reply by dzintars** [reply]
Other thing what I am thinking about is... to get things right, we need to look at bigger picture. And for me that picture is `Personal Knowledge Management (PKM)` and `Personal Development Environment (PDE)`. I think, it is worth to come up with some high level framework how all the pieces comes together to shape that environment. What APIs should be exposed, what integrations should be possible, etc.

**Reply by tbabej** [reply]
You might be interested in this somewhat popular vim plugin for taskwarrior that I wrote couple of years back :)

https://github.com/tools-life/taskwiki

**Comment by djmitche** [comment]
There are a few rust-tagged issues in this repo if you'd like to get your feet wet ;)
